rudder-lang/tools/cfjson_tester

```diff
@@ -14,70 +14,113 @@ import configparser
 
 def compare_json(first, second):
-  with open(first, "r") as f1:  
-    with open(second, "r") as f2:
-      j1 = json.load(f1)
-      j2 = json.load(f2)
+  try:
+    with open(first, "r") as f1:
+      try:
+        with open(second, "r") as f2:
+          j1 = json.load(f1)
+          j2 = json.load(f2)
+      except IOError as e:
+        print_json("compare_json", second, str(e), True)
+        exit(1)
+  except IOError as e:
+    print_json("compare_json", first, str(e), True)
+    exit(1)
   diff = jsondiff.diff(j1, j2, syntax="symmetric", marshal=True)
   ## diff hacks
   if 'bundle_name' in diff:
     if diff['bundle_name'][0] + "_technique" == diff['bundle_name'][1]:
       del diff['bundle_name']
-  ## TEST
+  ## print if diff
   if len(diff) != 0:
-    print(json.dumps(diff, indent=2, sort_keys=True))
+    print_json("compare_json", second, diff)
     exit(1)
 
 def load_cf(file_name):
   # Check file exists
   if not os.path.exists(file_name):
-    print("No such file: " + file_name)
+    print_json("compare_cf", file_name, 'No such file', True)
     exit(1)
 
   env = os.environ.copy()
   env['RES_OPTIONS'] = 'attempts:0'
-  out = check_output([ncf.CFENGINE_PATH, "-pjson", "-f", file_name], env=env)
+  out = check_output([ncf.CFENGINE_PATH, "-pjson", "-f", "{}/{}".format(os.getcwd(), file_name)], env=env)
   try:
     promises = json.loads(out)
   except Exception as e:
-    print("An error occured while parsing technique '"+file_name+"'")
+    print_json("compare_cf", file_name, 'An error occured while parsing technique: {}'.format(str(e)), True)
     exit(1)
   return promises
 
+def compare_cf(first, second):
+  j1 = load_cf(first)
+  j2 = load_cf(second)
+  diff = jsondiff.diff(j1, j2, syntax="symmetric", marshal=True)
+  remove_wilful_entries(diff)
+  empty_dict_cleaner(diff)
+  if diff and len(diff) != 0:
+    print_json("compare_cf", second, diff)
+
+
+## deletes from diff list every i/o difference done on purpose (ie that are not a bug)
+def remove_wilful_entries(diff):
+  ## del name since `_technique` is always added
+  name = deep_get(diff, ['bundles'], 'name')
+  if name[0] + '_technique' == name[1]:
+    deep_del(diff, ['bundles'], 'name', 'name')
+  ## del sourcePath since it is not relevant
+  deep_del(diff, ['bundles'], 'sourcePath', 'sourcePath')
+  ## del superfluous if true equivalent condition
+  deep_del(diff, ['bundles', 'promiseTypes', 'contexts', 'promises', 'attributes', 'rval', 'arguments'], {'type': 'string', 'value': 'any'}, '$delete')
+
+
 def empty_dict_cleaner(diff_dict):
   if type(diff_dict) is dict:
-    for element in diff_dict:
-      empty_dict_cleaner(diff_dict[element])
-      if len(diff_dict[element]) == 0:
+    for element in diff_dict.copy():
+      cur = diff_dict[element]
+      empty_dict_cleaner(cur)
+      if len(cur) == 0:
         diff_dict.pop(element)
         return
+      elif len(cur) == 1 and type(cur) is dict and 'line' in cur.keys():
+        diff_dict.pop(element)
```

Comment:
These two lines are the fix. 
The diff tool adds a lot of orphan lines dict, and when deleting "on purpose diffs" these "lines" remain and must be cleaned.

Reply:
the type(cur) check should come first since python does lazy evaluation and len isn't available for every ype

Reply:
ok fixing